### PR TITLE
Modified mpmc_queue member declaration order - issue 113

### PR DESCRIPTION
**Pull request by Drioastorga**
The mpmc queue class members have been reordered to fulfill the NL.16 GuideLine.

### CODE REVIEW

File: include/common/mpmc_queue.h [hunk 1]
```diff
@@ -30,28 +30,9 @@ constexpr int DEFAULT_SIZE = 100;
...
-      bool full(unsigned long long current);
-      bool empty(unsigned long long current);
-   public:  
+   public:
       typedef T value_type;
-      T pop();
-      bool push(T item);
-      bool empty();
-      
+
```
Comment (jdgarciauc3m) [inline include/common/mpmc_queue.h]: This typedef is private and not used. What is the point?

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
Revise which functions should be const.
Revise which functions can be safely made noexcept

File: include/common/mpmc_queue.h (hunk 1 above)
Comment (Drioastorga) [reply include/common/mpmc_queue.h]: This typedef is used to determine the type of the queue contents in some patterns.

Comment (jdgarciauc3m) [reply include/common/mpmc_queue.h]: There should be a better way.

Provide some pointer to example.

**Comment by Drioastorga** [comment]
The function members definitions have been accordingly modified to be const and noexcept.

File: include/common/mpmc_queue.h (hunk 1 above)
Comment (Drioastorga) [reply include/common/mpmc_queue.h]: typedef has been replaced to "using value_type = T"

File: include/common/mpmc_queue.h [hunk 2]
```diff
@@ -30,28 +30,9 @@ constexpr int DEFAULT_SIZE = 100;
...
 
-      bool full(unsigned long long current);
-      bool empty(unsigned long long current);
-   public:  
-      typedef T value_type;
-      T pop();
-      bool push(T item);
-      bool empty();
-      
       mpmc_queue<T>(int _size, bool active){
```
Comment (jdgarciauc3m) [inline include/common/mpmc_queue.h]: Move constructor out of class for consistent layout.

File: include/common/mpmc_queue.h [hunk 3]
```diff
@@ -70,6 +51,28 @@ class mpmc_queue{
...
+      int size;
+      std::vector<T> buffer;
+      std::atomic<unsigned long long> pread;
+      std::atomic<unsigned long long> pwrite;
+      std::atomic<unsigned long long> internal_pread;
+      std::atomic<unsigned long long> internal_pwrite;
+
+      bool lockfree = false;
+
+      std::mutex mutex;
```
Comment (jdgarciauc3m) [inline include/common/mpmc_queue.h]: A better name for mutex? It is a bad idea to give the same name to a variable than and std:: symbol

File: include/common/mpmc_queue.h [hunk 4]
```diff
@@ -70,6 +51,28 @@ class mpmc_queue{
...
+      std::condition_variable emptycv;
```
Comment (jdgarciauc3m) [inline include/common/mpmc_queue.h]: Better name `empty`

File: include/common/mpmc_queue.h [hunk 5]
```diff
@@ -70,6 +51,28 @@ class mpmc_queue{
...
+      std::condition_variable fullcv;
```
Comment (jdgarciauc3m) [inline include/common/mpmc_queue.h]: Better name: `full`

File: include/common/mpmc_queue.h [hunk 6]
```diff
@@ -70,6 +51,28 @@ class mpmc_queue{
          internal_pread = 0;
          internal_pwrite = 0;
       }
+
+      bool empty();
```
Comment (jdgarciauc3m) [inline include/common/mpmc_queue.h]: rename to `is_empty()`

File: include/common/mpmc_queue.h [hunk 7]
```diff
@@ -70,6 +51,28 @@ class mpmc_queue{
...
+      T pop();
+      bool push(T item);
+
+   private:
+      bool full(unsigned long long current);
```
Comment (jdgarciauc3m) [inline include/common/mpmc_queue.h]: rename to is_full()

File: include/common/mpmc_queue.h [hunk 8]
```diff
@@ -70,6 +51,28 @@ class mpmc_queue{
...
+      bool empty(unsigned long long current);
```
Comment (jdgarciauc3m) [inline include/common/mpmc_queue.h]: rename to is_empty()

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
More changes requested.

File: include/common/mpmc_queue.h (hunk 2 above)
Comment (Drioastorga) [reply include/common/mpmc_queue.h]: The constructor has been modified in issue-114 branch.

File: include/common/mpmc_queue.h (hunk 6 above)
Comment (Drioastorga) [reply include/common/mpmc_queue.h]: Renamed to is_empty()

File: include/common/mpmc_queue.h (hunk 8 above)
Comment (Drioastorga) [reply include/common/mpmc_queue.h]: Renamed to is_empty()

File: include/common/mpmc_queue.h (hunk 7 above)
Comment (Drioastorga) [reply include/common/mpmc_queue.h]: Renamed to is_full()

File: include/common/mpmc_queue.h (hunk 3 above)
Comment (Drioastorga) [reply include/common/mpmc_queue.h]: Rename to m

File: include/common/mpmc_queue.h (hunk 4 above)
Comment (Drioastorga) [reply include/common/mpmc_queue.h]: Renamed to empty

File: include/common/mpmc_queue.h (hunk 5 above)
Comment (Drioastorga) [reply include/common/mpmc_queue.h]: Renamed to full

**Review by jdgarciauc3m — APPROVED** [review]
OK

**Comment by jdgarciauc3m** [comment]
Solves issue #113

**Pull request merged**